File: apps/FaceEmotionRecognition/iOS/ZeticMLangeFaceEmotionRecognition-iOS/FaceEmotionRecognition.swift
import Foundation
import UIKit
import ZeticMLange

class FaceEmotionRecognition {
    private let model = (try? ZeticMLangeModel(personalKey: "YOUR_PERSONAL_KEY", name: "YOUR_PROJECT_NAME", version: 1))!
    private let wrapper = FaceEmotionRecognitionWrapper()
    
    func process(input: FaceEmotionRecognitionInput) -> FaceEmotionRecognitionOutput {
        do {
            let preprocess = wrapper.preprocess(input.image, input.roi)
            if (preprocess.isEmpty) {
                throw ZeticMLangeError("FaceEmotionRecognition", "preprocess failed")
            }
            try model.run([preprocess])
            var modelOutput = model.getOutputDataArray()
            let result = wrapper.postprocess(&modelOutput)
            return FaceEmotionRecognitionOutput(result: result)
        } catch {
            print(error)
        }
        return FaceEmotionRecognitionOutput(result: FaceEmotionRecognitionResult(emotion: "", confidence: 0))
    }
}

struct FaceEmotionRecognitionInput: AsyncFeatureInput {
    let image: UIImage
    let roi: Box
}

struct FaceEmotionRecognitionOutput: AsyncFeatureOutput {
    let result: FaceEmotionRecognitionResult
    
    init(result: FaceEmotionRecognitionResult = FaceEmotionRecognitionResult(emotion: "", confidence: 0)) {
        self.result = result
    }
}

File: apps/FaceEmotionRecognition/Android/app/src/main/java/com/zeticai/faceemotionrecognition/feature/FaceEmotionRecognition.kt
package com.zeticai.faceemotionrecognition.feature

import android.content.Context
import com.zeticai.mlange.core.model.ZeticMLangeModel
import com.zeticai.mlange.feature.entity.Box
import com.zeticai.mlange.feature.faceemotionrecognition.FaceEmotionRecognitionResult
import com.zeticai.mlange.feature.faceemotionrecognition.FaceEmotionRecognitionWrapper

class FaceEmotionRecognition(
    context: Context,
) {
    private val model: ZeticMLangeModel = ZeticMLangeModel(context, "YOUR_PERSONAL_KEY", "YOUR_PROJECT_NAME")
    private val wrapper: FaceEmotionRecognitionWrapper = FaceEmotionRecognitionWrapper()

    fun run(imagePtr: Long, roi: Box): FaceEmotionRecognitionResult {
        val preprocess = wrapper.preprocess(imagePtr, roi)
        model.run(arrayOf(preprocess))
        val modelOutput = model.outputBuffers
        return wrapper.postprocess(modelOutput.map {
            val byteArray = ByteArray(it.remaining())
            it.get(byteArray)
            return@map byteArray
        }.toTypedArray())
    }

    fun close() {
        model.deinit()
        wrapper.deinit()
    }

    companion object {
        val modelKey: String = "223fed6191c848df8b2b707b76707baa"
    }
}